### PR TITLE
Ft user registration 187300143

### PR DESCRIPTION
**Pull request by TuyishimireEric**
#### What does this PR do?

User Registration

#### Description of Task to be completed?

- Users once registered via the registration endpoint, should receive a JWT to be required on all subsequent calls to all other endpoints that require authentication.

Users should receive a descriptive error message when they enter invalid user parameters in an attempt to signup.

#### How should this be manually tested?

- run `npm install` 
- run `npm run migrate`
- run `npm run seed`
- run `npm run dev`
- check the [swagger documentation ](http://localhost:3000/swagger/#/users/post_api_users_register)

#### Any background context you want to provide?

-dummy data are provided before having the real data

#### What are the relevant pivotal tracker/Trello stories?

- id of the project
#187300143

### CODE REVIEW

**Review by Ruthuwamahoro — APPROVED** [review]
Hello! Ruth here. congratulations! You did an amazing job

**Review by manzidavidrw — APPROVED** [review]

**Review by Nadinefiona — APPROVED** [review]

**Review by RWEMAREMY — APPROVED** [review]
Good work

File: .eslintrc.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
       "import/no-unresolved": [2, { "commonjs": true }],
       "no-shadow": ["error", { "allow": ["req", "res", "err"] }],
       "valid-jsdoc": [
-        "error",
+        "off",
```
Comment (fab-ryan) [inline .eslintrc.json]: why did you remove this it's needed and jsondoc where needed or ingore in file but not entire projects

File: README.md [hunk 2]
```diff
@@ -1,3 +1,117 @@
...
+
+# 📖 E-COMMERCE<a name="about-project"></a>
+
+This repository contains the backend for an e-commerce application built with Node.js, Express, and PostgreSQL. It includes a comprehensive suite of tests using Mocha to ensure the reliability and functionality of the application.
+
+# 🧾 Key Features <a name="key-features"></a>
+ The backend provides the following features:
+
+### Basic Information: 
+Users can register and login to the system
```
Comment (fab-ryan) [inline README.md]: Good work

File: combined.log
Comment (fab-ryan) [inline combined.log:1]: remove logs or add it in gitingore

File: src/database/db.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@
 import { Sequelize } from 'sequelize';
 import getDatabaseConfig from '../config/config';
 import Models from './models';
+import logger from '../utils/logger';
```
Comment (fab-ryan) [inline src/database/db.ts]: can you make folder structure like this
- if it's utils create new file index where it's will export every file from utils folder then all you will need to do is to import ../utils only that's enaught and simple

File: src/middleware/validateUser.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    return sendResponse<null>(res, 400, null, error.message);
+  }
+
+  const passwordError = testPassword(req.body.password);
+  if (passwordError) {
+    return sendResponse<null>(res, 400, null, passwordError);
+  }
+
+  next();
+};
```
Comment (fab-ryan) [inline src/middleware/validateUser.ts]: also you and @Ruthuwamahoro you have to harmonize this help function where it will be used by everyone if you need help from this I can help you tomorrow make it resuable

File: src/utils/logger.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    }),
+    format.errors({ stack: true }),
+    format.prettyPrint(),
+    format.json()
+  ),
+  defaultMeta: { service: 'alpha-ec-be' },
+  transports: [
+    new transports.File({ filename: 'error.log', level: 'error' }),
+    new transports.File({ filename: 'combined.log' }),
+  ],
```
Comment (fab-ryan) [inline src/utils/logger.ts:25]: add those loggs  into one folder

File: src/utils/response.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+): Response => {
+  const result = /^20\d$/.test(status.toString()) ? 'Success!' : 'Error!';
+  return res.status(status).json({
+    status: result,
+    data,
+    message,
+  });
+};
+
+export default sendResponse;
```
Comment (fab-ryan) [inline src/utils/response.ts]: good work @TuyishimireEric on this helper

**Review by fab-ryan — CHANGES_REQUESTED** [review]
work on those comments

File: README.md (hunk 2 above)
Comment (TuyishimireEric) [reply README.md]: Thank you 🙏

File: .eslintrc.json (hunk 1 above)
Comment (TuyishimireEric) [reply .eslintrc.json]: Resolved

File: src/utils/response.ts (hunk 6 above)
Comment (TuyishimireEric) [reply src/utils/response.ts]: Thank you 🙏

File: src/utils/logger.ts (hunk 5 above)
Comment (TuyishimireEric) [reply src/utils/logger.ts:25]: I have changed this function to check if the `logs` is not available and creates it to keep all the logs  👍

File: src/middleware/validateUser.ts (hunk 4 above)
Comment (TuyishimireEric) [reply src/middleware/validateUser.ts]: I have refactored this function

File: src/database/db.ts (hunk 3 above)
Comment (TuyishimireEric) [reply src/database/db.ts]: Resolved

File: src/utils/validateNewUser.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    return error.message;
+ }
+
+ const passwordError = testPassword(body.password);
+ if (passwordError) {
+    return passwordError;
+ }
+
+ return null;
+};
```
Comment (fab-ryan) [inline src/utils/validateNewUser.ts]: still you haven't optimized this function you can call it validations let's me give sample implementation how it's should work:
``export const validate =
  (schema: typeof ObjectSchema, type: requestType)=>(req,res,_:NextFunction)=> ``
  the request type can be body, params or query,
  name those variable as you want

File: src/middleware/isNewUserValid.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+export const isNewUserValid = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const validationError = validateNewUser(req.body);
+
+  if (validationError) {
+    return sendResponse<null>(res, 400, null, validationError);
+  }
```
Comment (fab-ryan) [inline src/middleware/isNewUserValid.ts]: then make this as also middlware

**Review by fab-ryan — COMMENTED** [review]
Make those changes

File: package.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
   "description": "alpha team project",
   "main": "index.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=test nyc --reporter=text --reporter=lcov mocha  --require ts-node/register   test/server-run.ts --coverage",
+    "test": "cross-env NODE_ENV=test nyc --reporter=text --reporter=lcov mocha  --require ts-node/register   test/*.ts --coverage",
```
Comment (fab-ryan) [inline package.json]: I have to ask who the test will run without migration or demigration think about this

File: src/middleware/validate.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+) => {
+
+  try {
+    const { error } = schema.validate(req.body, { abortEarly: false })
+
+    if(error){
+      const context = (error.details[0] as any).context.message;
+      const errorMessage = context? context : error.message;
+      
+      console.log(error);
```
Comment (fab-ryan) [inline src/middleware/validate.ts]: here you will return only one message but when we have alot of errors what will happened?? and remember to remove console.log

File: src/routers/index.ts [hunk 11]
```diff
@@ -1,9 +1,8 @@
 import { Router, Request, Response } from 'express';
-import DummyRouter from './dummyRoute';
+import UserRouter from './userRoute';
 
 const router = Router();
-const routers: Router[] = [DummyRouter];
-router.use('/api', routers);
+router.use('/api/users', UserRouter);
```
Comment (fab-ryan) [inline src/routers/index.ts]: can you format this like 

 `const routers: Router[] = [ UserRouter ]`
 `router.use('/api', routers);`
 this will help when every router will be put under that array not creatng new router, I know someone may forgett to add api and start complaining

File: test/user.test.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+           email: 'test@example.com',
+           phone: '1234567890',
+           address: '123 Test Street',
+           password: '11111@aa'
+        };
+       
+        chai
+           .request(app)
+           .post('/api/users/register')
+           .send(user)
```
Comment (fab-ryan) [inline test/user.test.ts]: can you make global setup for this if it's not complicated

**Review by fab-ryan — CHANGES_REQUESTED** [review]
work on those comment and let me know when you are done

File: src/routers/index.ts (hunk 11 above)
Comment (TuyishimireEric) [reply src/routers/index.ts]: Done 👍

File: src/middleware/validate.ts (hunk 10 above)
Comment (TuyishimireEric) [reply src/middleware/validate.ts]: Done 👍

File: README.md [hunk 13]
```diff
@@ -1,5 +1,5 @@
-# al# alpha-ec-be
-[![Coverage Status](https://coveralls.io/repos/github/atlp-rwanda/alpha-ec-be/badge.svg)](https://coveralls.io/github/atlp-rwanda/alpha-ec-be)
+
+[![Coverage Status](https://coveralls.io/repos/github/atlp-rwanda/alpha-ec-be/badge.svg?branch=develop)](https://coveralls.io/github/atlp-rwanda/alpha-ec-be?branch=develop)
```
Comment (fab-ryan) [inline README.md]: here use this 
`[Coverage Status](https://coveralls.io/github/atlp-rwanda/alpha-ec-be)`

 insteady

File: src/controllers/userController.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      address,
+      password: hashedPassword,
+    });
+
+    await user.save();
+
+    const token = jwt.sign(
+      { id: user.id },
+      process.env.JWT_SECRET || 'secret',
+      { expiresIn: '2h' }
```
Comment (fab-ryan) [inline src/controllers/userController.ts]: can you add this secret in config.js because we will use it in reset password to send token

**Review by fab-ryan — COMMENTED** [review]
those two things only then I can merge this PR

File: src/controllers/userController.ts (hunk 14 above)
Comment (TuyishimireEric) [reply src/controllers/userController.ts]: the default 'secret'?

**Review by fab-ryan — APPROVED** [review]
